### PR TITLE
ARROW-16005: [Java] Fix ArrayConsumer when using ArrowVectorIterator

### PR DESCRIPTION
**Pull request by tom-s-powell**
Fixes https://issues.apache.org/jira/browse/ARROW-16005.

### CODE REVIEW

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java [hunk 1]
```diff
@@ -64,6 +63,14 @@ public void close() throws Exception {
     this.delegate.close();
   }
 
+  @Override
+  public void resetValueVector(ListVector vector) {
+    super.resetValueVector(vector);
+
+    FieldVector childVector = vector.getDataVector();
```
Comment (tom-s-powell) [inline java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java:70]: When `VectorSchemaRoot` is reused in `ArrowVectorIterator`, we currently hit the issue that the `currentIndex` here is reset to `0` but is never updated in the delegate consumer. As such, subsequent iterations will result in `null` array values because the `ListVector` (and data vector) is reset https://github.com/apache/arrow/pull/12692/files#diff-f812c76a565e7c56500943f512b8498487209b15ed036d404d703854841df3d0R150. 

For example, if you have a batch size of 2 and a `ResultSet` with 4 rows, the second iteration will be writing values into index 0 and 1 in the `ListVector` but the offsets for those in the data vector will be pointing at null values (because it was reset) and the values written to the data vector will be at larger indexes.

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java [hunk 2]
```diff
@@ -134,11 +122,26 @@ private VectorSchemaRoot createVectorSchemaRoot() {
...
       throw new RuntimeException("Error occurred while creating schema root.", e);
     }
+
+    // ensure consumers have been initialized
+    ensureInitialized(root);
     return root;
   }
 
+  private void ensureInitialized(VectorSchemaRoot root) throws SQLException {
+    if (!initialized) {
```
Comment (lidavidm) [inline java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java:132]: It seems this isn't right if `!config.isReuseVectorSchemaRoot()` because we'll have recreated the root. Since `ensureInitialized` is only ever called when creating a new root, I don't think we need to guard this with `initialized`?

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java [hunk 3]
```diff
@@ -90,13 +97,12 @@ public void consume(ResultSet resultSet) throws SQLException, IOException {
         int count = 0;
         try (ResultSet rs = array.getResultSet()) {
           while (rs.next()) {
-            ensureInnerVectorCapacity(innerVectorIndex + count + 1);
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java:93]: It's because of how a ListVector is laid out in memory. The list `[[1, 2], [], [3, 4, 5]]` is represented as the child vector `[1, 2, 3, 4, 5]` and the offsets `[0, 2, 2, 5]`. `ensureInnerVectorCapacity` is resizing the child vector, so when we call `consume` for the last element, we want to ensure the child vector has enough capacity for the current elements, along with all the previous elements, and it looks like that's what `innerVectorIndex` is tracking. 

In other words when we call `consume` for what will be `[3, 4, 5]` we need to ensure the child vector has space for at least 3, 4, 5, … elements not 1, 2, 3… elements.

**Review by lidavidm — COMMENTED** [review]
Thanks for fixing this!

Is there a unit test we can add to cover this?

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java (hunk 2 above)
Comment (tom-s-powell) [reply java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java]: Ah yes good point!

Comment (tom-s-powell) [reply java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java:141]: Actually, it's fine as its currently written. After creating a new `VectorSchemaRoot` we'll call `load` and this will call `JdbcConsumer#resetValueVector` passing the vectors of the _new_ `VectorSchemaRoot` and obtaining a reference to them (https://github.com/apache/arrow/pull/12692/files#diff-f812c76a565e7c56500943f512b8498487209b15ed036d404d703854841df3d0R161). This will happen before we consume data.

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lidavidm) [reply java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java:93]: I'm still not entirely convinced we can remove `innerVectorIndex` here?

**Review by lidavidm — COMMENTED** [review]
Thanks for plugging away at this. As mentioned it would be good to see a unit test to cover this too (I guess with both reusing/not reusing the VectorSchemaRoot)

@toddfarmer would you like to take a glance here as well?

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java (hunk 2 above)
Comment (tom-s-powell) [reply java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/ArrowVectorIterator.java:141]: Actually sorry I see we have to recreate the delegate consumer as it references the child vector so it does make sense to initialize each time.

**Comment by tom-s-powell** [comment]
I've updated the unit tests for `arrow-jdbc` to include array types.

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tom-s-powell) [reply java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/ArrayConsumer.java]: I've reverted the change to `innerVectorIndex`. Only thing I've done is reset it to 0 when we reset inner vector.

I've added unit tests for Arrays which covers the two issues when reusing `VectorSchemaRoot`:
1) NPE error from `ArrowVectorIterator`.
2) `ArrayConsumer` not reseting the delegate consumer.

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/Table.java [hunk 4]
```diff
@@ -204,6 +219,11 @@ public void setRowCount(int rowCount) {
     this.rowCount = rowCount;
   }
 
+  @Override
+  public String toString() {
+    return "Table{name='" + name + "', type='" + type + "'}";
```
Comment (tom-s-powell) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/Table.java:224]: Have added this for debugging purposes, helpful in the `ParameterizedTest` to easily see which test file is failing. I updated the table names to match the YML file name.

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowVectorIteratorTest.java [hunk 5]
```diff
@@ -99,15 +104,25 @@ public void testJdbcToArrowValues() throws SQLException, IOException {
...
   @Test
   public void testVectorSchemaRootReuse() throws SQLException, IOException {
-    String[] expectedColValues = {
+    String[] expectedIntColValues = {
         "[101, 102, 103]",
         "[104, null, null]",
         "[107, 108, 109]",
         "[110]"
     };
+    String[] expectedArrayColValues = {
```
Comment (tom-s-powell) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowVectorIteratorTest.java:113]: Added this test case for arrays specifically to test the fix to `ArrayConsumer` when resetting the delegate. Without the reset you end up reading null values.

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/JdbcToArrowTestHelper.java [hunk 6]
```diff
@@ -378,4 +393,25 @@ public static String[] getValues(String[] values, String dataType) {
...
+        String[] row = data.replace("(", "").replace(")", "").split(";");
+        Integer[] arr = new Integer[row.length];
+        for (int j = 0; j < arr.length; j++) {
+          arr[j] = "null".equals(row[j]) ? null : Integer.parseInt(row[j]);
+        }
+        valueArr[i++] = arr;
+      }
+    }
+    return valueArr;
+  }
```
Comment (lidavidm) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/JdbcToArrowTestHelper.java:416]: It feels like this could be reused from Table.getListValues? Especially as this seems to also handle nulls

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowVectorIteratorTest.java [hunk 7]
```diff
@@ -119,7 +134,10 @@ public void testVectorSchemaRootReuse() throws SQLException, IOException {
       assertNotNull(cur);
 
       // verify the first column, with may contain nulls.
-      assertEquals(expectedColValues[batchCount], cur.getVector(0).toString());
+      assertEquals(expectedIntColValues[batchCount], cur.getVector(0).toString());
```
Comment (lidavidm) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowVectorIteratorTest.java:137]: Is there a better way to assert equality than the string representation (perhaps with getObject as done elsewhere)? This is a little brittle

**Review by lidavidm — APPROVED** [review]
Sorry for the delay.

Thanks for adding the tests and refactoring everything! I left some minor questions but I think this is ready.

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowVectorIteratorTest.java (hunk 7 above)
Comment (tom-s-powell) [reply java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/h2/JdbcToArrowVectorIteratorTest.java]: Switched to use the `assertIntVectorValues` and `assertListVectorValues`

**Review by lidavidm — APPROVED** [review]
